# DTSC 580: Data Manipulation
## Assignment: LEGO Project
### Name: 

![LegoSchema](Lego_Schema.PNG)

## Overview

In this assignment, you will load the various tables from the [Rebrickable](https://rebrickable.com/) website and perform analysis on the data.  Rebrickable is a fantastic website that is used to show you which LEGO sets you can build from the sets and parts you already own.  They will even show you which parts you are missing to complete the new set and suggest sets that you could purchase to get those parts.  

## Data

The data has been downloaded directly from their website and includes files for the tables shown in the schema above.  Your first task will be to go through the data and review the schema to understand how the tables all fit together.  There is no data dictionary that I am aware of, but you should be able to understand this data by reviewing the files and schema and visiting their [Download](https://rebrickable.com/downloads/) page and [API](https://rebrickable.com/api/) documentation.  **Note that you must use the files downloaded from Brightspace for this assignment and not download new files from the Rebrickable website.**

## Assignment

In this assignment, you will create a number of functions to complete certain tasks.  This is not needed if you were working on this project on your own, but since we want to see if you can select the data via Pandas/Numpy, we will use those functions in the automatic grading in CodeGrade.  Without this, students could try to look up the answers in the `csv` files instead of wrangling the data with Python/Pandas, which would defeat the purpose of this class.

In addition, we would strongly suggest that you merge data sets as appropriate for the given task.  For example, you might merge the `sets` and `themes` DataFrames to answer one question, and then merge `sets`, `inventories`, `inventory_minifigs`, and `minifigs` to answer another.  This will make your code much easier to write and will be key in working with data across multiple tables.  In fact, I was able to create each function with only one line of code because of the merging of certain data sets.  Also note that you may not use all the tables in this assignment.  This is also an important skill for a Data Scientist to know when to use data and what data is unnecessary.

When complete, save your notebook as `lego.ipynb` and submit to CodeGrade for automatic grading.

## Note

<u>Show Work</u>

Remember that you must show your work.  Students submissions are spot checked manually throughout the term to verify that they are not hard coding the answer from looking only in the file or in CodeGrade's expected output.  If this is seen, the student's answer will be manually marked wrong and their grade will be changed to reflect this. 

For example, if the answer to Q1, the mean of a specific column, is 22:
```
# correct way
Q1 = df['column_name'].mean()

# incorrect way
Q1 = 22 
```

Let's start out by importing some standard imports.

In [1]:
# common imports
import pandas as pd
import numpy as np

# Do not change this option; This allows the CodeGrade auto grading to function correctly
pd.set_option('display.max_columns', None)

## Import Data

Load the various Lego datasets below calling them the same as the name of the table as seen in the schema.  Note that you may or may not have to use all the data sets.  One of the things that all Data Scientists and Analysts have to decide when working on a project is what data is useful for a task and what data is not.  Remember to look at the schema above to see how all the data sets are connected.

The autograder in CodeGrade will have the `csv` files in the same location where the notebook is run.  Because of this, please make sure your local files are in the same location as your notebook.

In [2]:
# import Lego data sets here
# make sure that files are in the same folder as your notebook
inventories = pd.read_csv('inventories.csv')
inventory_sets = pd.read_csv('inventory_sets.csv')
sets = pd.read_csv('sets.csv')
themes = pd.read_csv('themes.csv')
inventory_minifigs = pd.read_csv('inventory_minifigs.csv')
minifigs = pd.read_csv('minifigs.csv')
inventory_parts = pd.read_csv('inventory_parts.csv')
part_categories = pd.read_csv('part_categories.csv')
parts = pd.read_csv('parts.csv')
colors = pd.read_csv('colors.csv')
part_relationships = pd.read_csv('part_relationships.csv')
elements = pd.read_csv('elements.csv')

## Exploration

Take time to look at all of the data sets using `.head()`, `.info()`, `.describe()`, etc to familiarize yourself with the data sets.  As you look at them, notice how the various tables connect together.

In [23]:
sets.head()

set_num                        name  year  theme_id  num_parts
0       001-1                       Gears  1965         1         43
1      0011-2           Town Mini-Figures  1979        67         12
2      0011-3  Castle 2 for 1 Bonus Offer  1987       199          0
3      0012-1          Space Mini-Figures  1979       143         12
4      0013-1          Space Mini-Figures  1979       143         12
..        ...                         ...   ...       ...        ...
995  122005-1           Tyrannosaurus Rex  2020       602         65
996  122006-1                 Triceratops  2020       602         65
997  122007-1             Owen with Canoe  2020       602         17
998  122008-1                    Dino Lab  2020       602         28
999  122009-1      Vic Hoskins with Buggy  2020       602         35

[1000 rows x 5 columns]

## Assignment

Answer all the questions below using the available data.  There are multiple ways to get some of the answers.  I would suggest that you attempt to merge some data sets as needed to assist with getting answers.  This will make your tasks a lot easier and your code will be cleaner.

Also, you will be creating functions to get the answers to some of these questions.  This will make it much easier for you to get answers to similar questions in the future and CodeGrade will be checking your functions using different inputs.  

**Q1:** Create a function called `data_count` that takes as input the data set (for example: `sets`) and returns the total number of rows in the data (as an integer).  Nothing else should be returned.

In [3]:
def data_count(data):
    return len(data)

**Code Check:**  If you call your `data_count` function using the `sets` DataFrame, you should see that there are 18,576 sets represented in the data.

In [4]:
data_count(sets)

18576

**Q2:** What is the average number of parts for a Lego set (rounded to the nearest integer)?  Save this as `Q2`.

In [11]:
Q2= int(sets['num_parts'].mean().round())
Q2

161

**Q3:** What is the median number of parts per set (output as an integer)?  Save this as `Q3`.

In [12]:
Q3 = int(sets['num_parts'].median())
Q3

37

You should see that the median number of parts is much lower than the average number of parts per set.  This tells us that there are some Lego sets with a very high number of parts that brings the average up.

**Q4:** What is the largest number of parts that a set has? Save this as `Q4`.

In [14]:
Q4 = sets['num_parts'].max()
Q4

11695

**Q5:** Given the set with the largest number of parts calculated above, select its row from the `sets` DataFrame.  Save this as `Q5`.  Do not reset the index.

In [15]:
Q5 = sets[sets['num_parts'] == Q4]
Q5

set_num       name  year  theme_id  num_parts
3867  31203-1  World Map  2021       709      11695

**Q6:** Create a function called `select_set_row` where you pass it the `set_num` (as a string) and the function returns the respective row of the DataFrame from `sets`. Do not reset the index.

In [16]:
def select_set_row(set_num):
    return sets[sets['set_num'] == set_num]

select_set_row('001-1')

set_num   name  year  theme_id  num_parts
0   001-1  Gears  1965         1         43

**Q7:** Create a function called `select_set_numparts` that takes as input a specific number of parts (as an integer) and selects all the rows in the `sets` DataFrame that contain those number of parts.  Return the DataFrame sorted by `year` and then by `set_num`.  Do not reset the index.

In [28]:
def select_set_numparts(num_parts):
    return sets[sets['num_parts'] == num_parts].sort_values(by=['year','set_num'])

select_set_numparts(1)

set_num                                         name  year  theme_id  \
11059  700.1.1-1                     Single 2 x 4 Brick (ABB)  1950       371   
11062  700.1.2-1                     Single 2 x 2 Brick (ABB)  1950       371   
11082  700.B.1-1  Single 1 x 4 x 2 Window without Glass (ABB)  1950       371   
11083  700.B.2-1  Single 1 x 2 x 3 Window without Glass (ABB)  1950       371   
11084  700.B.3-1  Single 1 x 2 x 2 Window without Glass (ABB)  1950       371   
...          ...                                          ...   ...       ...   
10718    66666-1                         LEGO Masters Co-pack  2021       621   
15589   854084-1                 2x4 White Metallic Key Chain  2021       503   
15963    88015-1                                  Battery Box  2021       655   
15965    88017-1                          Large Angular Motor  2021       655   
15966    88018-1                         Medium Angular Motor  2021       655   

       num_parts  
11059          1  
11062          1  
11082          1  
11083          1  
11084          1  
...          ...  
10718          1  
15589          1  
15963          1  
15965          1  
15966          1  

[441 rows x 5 columns]

**Code Check:** Call the `select_set_numparts()` function to select all sets with `0` number of parts.  There should be 3,146 rows returned.

In [29]:
select_set_numparts(0)

set_num                                               name  \
11061          700.1-2                     Extra-Large Gift Set (Mursten)   
17563         b55dk-01               Lego Mursten - System i Leg Byggebog   
17564         b56de-01                                    System im Spiel   
11072          700.2-2                                       Gift Package   
11078         700.3A-2                                       Gift Package   
...                ...                                                ...   
15619         854195-1                                    Ernie Key Chain   
15620         854197-1                               Demogorgon Key Chain   
17170  9780744054583-1  DC Comics Super Heroes: Character Encyclopedia...   
17193  9780794447557-1                               City: Color The City   
17200  9780794448172-1           DC Comics Super Heroes: Ready For Action   

       year  theme_id  num_parts  
11061  1953       366          0  
17563  1955       497          0  
17564  1956       497          0  
11072  1957       366          0  
11078  1957       372          0  
...     ...       ...        ...  
15619  2022       503          0  
15620  2022       503          0  
17170  2022       497          0  
17193  2022       497          0  
17200  2022       497          0  

[3146 rows x 5 columns]

**Q8:** 
- The oldest set listed was made in what year?  Save the year as `Q8A` (as an integer).  
- The newest set listed was made in what year? Save the year as `Q8B` (as an integer).

In [33]:
#The oldest set listed was made in what year?  Save the year as `Q8A`
Q8A = int(sets['year'].min())
type(Q8A)

int

In [34]:
Q8B = int(sets['year'].max())
Q8B

2022

**Q9:** 
- Create a function called `select_set_year` that takes as input a year (as an integer) and returns the rows from the `sets` DataFrame that matches that year.  Sort the DataFrame by `set_num`.  Do not reset the index.
- Use the above function to return a DataFrame for rows that match the minimum year (calculated as `Q8A` above) and call this `Q9A`.  
- Then use the above function to return a DataFrame for rows that match the maximum year (calculated as `Q8B` above) and call this `Q9B`.

In [39]:
def select_set_year(year):
    return sets[sets['year'] == year].sort_values(by=['set_num'])

Q9A = select_set_year(Q8A)	
Q9A

set_num                               name  year  theme_id  num_parts
11057  700.1-1         Extra-Large Gift Set (ABB)  1949       365        142
11070  700.2-1               Large Gift Set (ABB)  1949       365        178
11075  700.3-1              Medium Gift Set (ABB)  1949       365        142
11080  700.A-1              Small Brick Set (ABB)  1949       371         24
11081  700.B-1  Small Doors and Windows Set (ABB)  1949       371         12

In [38]:
Q9B = select_set_year(Q8B)
Q9B    

set_num                                               name  \
336            10297-1                                     Boutique Hotel   
600            10781-1                          Spider-Man’s Techno Trike   
601            10782-1                      Hulk vs. Rhino Truck Showdown   
602            10783-1                        Spider-Man at Doc Ock’s Lab   
603            10784-1                     Spider-Man Webquarters Hangout   
...                ...                                                ...   
17170  9780744054583-1  DC Comics Super Heroes: Character Encyclopedia...   
17193  9780794447557-1                               City: Color The City   
17200  9780794448172-1           DC Comics Super Heroes: Ready For Action   
17207  9780794449247-1  DC Comics Super Heroes: Gotham City’s New Defe...   
17726    DOTS-SERIES-6                    DOTS Collectible Tiles Series 6   

       year  theme_id  num_parts  
336    2022       155       3068  
600    2022       706         59  
601    2022       706        110  
602    2022       706        131  
603    2022       706        155  
...     ...       ...        ...  
17170  2022       497          0  
17193  2022       497          0  
17200  2022       497          0  
17207  2022       497          7  
17726  2022       688         16  

[183 rows x 5 columns]

**Q10:** What is the median year represented in the `sets` data?  Save this as `Q10` (as an integer).

In [40]:
Q10 = int(sets['year'].median())
Q10

2010

Notice that the first set was back in 1949 but over half of the sets were after the median year. This could be an indication that the number of sets has increased greatly over the past number of years.

**Q11:** What is the average number of sets listed per year?  Save this as `Q11` (as an integer).

In [41]:
Q11 = int(sets.groupby('year').size().mean())
Q11

258

**Q12:** 
- Create a function called `theme_by_year` that takes as input a year (as an integer) and shows the theme ids and theme names (listed in order by theme id) that were in sets that year.  
- The column names must be `id` and `name_themes` (to differentiate between the name of a theme and the name of a set) in that order.  
- The index should be reset and go from `0` to `n-1`.  
- Each theme should only be listed once even if it appeared in more than one set from that year -- duplicate themes should be based on theme id and not name since there are some themes with the same name but with a different id.
- Hint:  It will help if you were to think about merging appropriate DataFrames to help you get this answer.

In [48]:
def theme_by_year(year):
    return themes.merge(sets[sets['year'] == year], on='theme_id').drop_duplicates(subset='theme_id').sort_values(by=['theme_id']).reset_index(drop=True)[['theme_id','name_themes']]
theme_by_year(2010)

KeyError: 'theme_id'

**Code Check:** Call the `theme_by_year()` function on the year `1960`.  Your output should look as follows:

| | id | name_themes|
|-|---|---|
|0|371|Supplemental|
|1|497|Books|
|2|513|Classic|

**Optional Practice:** Call the `theme_by_year()` function using your birth year to see the Lego themes that were popular when you were born.

In [47]:
# Call the `theme_by_year()` function on the year `1960`
Q13 = theme_by_year(1960)

KeyError: 'theme_id'

**Q13:** What theme has the most number of sets associated with it?  Save the theme name as `Q13` (as a string).

In [49]:
# What theme has the most number of sets associated with it?  Save the theme name as `Q13` (as a string).
Q14 = themes[themes['theme_id'] == Q13['theme_id'].max()]['name_themes'].values[0]
Q14

KeyError: 'theme_id'

**Q14:** 
- Create a function called `theme_by_name` that takes as input a theme name and outputs all the sets (in `set_num` order) that had that theme.  
- The DataFrame should include the following columns in this order and with these exact names: `set_num`,`name_sets`,`year`,`num_parts`, `theme_id`, and `name_themes`.  
- The index should go from `0` to `n-1`.

In [ ]:
### ENTER CODE HERE ###

**Code Check:** Call the `theme_by_name()` function using the theme name from `Q13`.  You should return 1,904 sets that match that theme name.

In [ ]:
### CODE CHECK ###

**Q15:** Create a function called `theme_by_setnum` that takes as input the set_num (as a string) and returns the respective theme name as a string.

In [ ]:
### ENTER CODE HERE ###

**Code Check:** Visit the [Sets page](https://rebrickable.com/sets/) from the Rebrickable website and find a set that you like.  Call your `theme_by_setnum()` function on that set and make sure the correct theme name is output.

In [ ]:
### CODE CHECK ###

**Q16:** Find all the sets named 'Millennium Falcon'. Of these sets, find the set with the largest number of parts.  You **do not** need to save this as `Q16`.  Instead you will use this information for later questions.

In [ ]:
### ENTER CODE HERE ###

**Q17:** 
- How many minifigs are contained in the set from `Q16`?  Save this as `Q17` (as an integer).  
- Note that there are sometimes multiple versions of a set as shown in the `inventories` table.  This will not matter for this question as there is only one version for the respective set but this should be noted for future questions.
- Remember to show your work or your answer will manually get marked as incorrect.

In [ ]:
### ENTER CODE HERE ###

**Q18:** 
- Create a function called `get_minifigs` that takes as input a set number (as a string) and a version number (as an integer) and outputs all minifigs from that set / version combination.  
- The DataFrame should be in `fig_num` order with the following columns/column names: `set_num`, `fig_num`, `fig_name`, `fig_num_parts`.  
- The index should go from 0 to n-1.

In [ ]:
### ENTER CODE HERE ###

**Q19:** Using the above function, output information about the minifigs from the set discussed in `Q16`.  Save this output as `Q19`.

In [ ]:
### ENTER CODE HERE ###

**Code Check:** Make sure that your `get_minifigs()` function works with both version 1 and version 2 of set 10217-1, Harry Potter's Diagon Alley.  Both versions should have 11 minifigs in the DataFrame.

In [ ]:
### CODE CHECK ###

**Q20:** 
- Create a function called `minifigs_from_themes` that takes as input a theme id (as an integer) and outputs all the minifigs contained in sets with that theme.  
- The DataFrame should be in `fig_num` order with the following columns/column names: `fig_num`, `fig_name`. 
- The index should go from 0 to n-1.  
- Exact duplicates should be removed so that only unique figs are included.

In [ ]:
### ENTER CODE HERE ###

**Code Check:** Call the `minifigs_from_themes()` function using the theme number of the set discussed in `Q16`.  There should be 149 minifigs output for that respective theme.

In [ ]:
### CODE CHECK ###

**Q21:** 
- Create a function called `sets_from_minifig` that takes as input a minifig number (as a string) and returns a list of sets a minifig has appeared in.  
- Return a DataFrame (sorted by `set_num` and then by `version`) with the following columns: `set_num`, `version`,`name_sets`, `fig_num`, `fig_name`.  
- The index should go from 0 to n-1.

In [ ]:
### ENTER CODE HERE ###     

**Code Check:** Call your `sets_from_minifig()` function on one of the Lego Friends minifigs, Andrea, who has figure number: `fig-001039`.  Your function should return a DataFrame that contains the following sets:

- Andrea on the Beach
- Birthday Party
- City Park Cafe

In [ ]:
### CODE CHECK ###

**Q22:** 

*Note:  Before you begin working on this question, please note that CodeGrade has a 1GB RAM limit.  If you see an `Exit -9` code when submitting your assignment, it is probably due to this RAM limit and previous students had an issue with this question causing too much memory usage.  The biggest thing to think about is to only use the tables that you need to answer for this question.  For instance, you do not need any of the minifigs data since this is not needed for the output and the merging of this data could cause the number of rows to "explode" in terms of the total count.  I understand that this can be annoying that we have to work around CodeGrade's limitations, but it is good practice in memory management.*

- Create a function called `set_parts` that takes as input the set number (as a string) and version number (as an integer) and outputs a DataFrame with each part in the set.  
- The DataFrame should have the following columns/column names: `set_num`, `version`, `name_sets`, `year`, `name_themes`, `part_num`, `name_part`, `name_color`,`quantity`,`is_spare`.  
- The Data should be sorted by `part_num` and then `name_color`.  
- The index should go from 0 to n-1.

In [ ]:
### ENTER CODE HERE ###

**Code Check**: Using the `set_parts()` function, output a list of parts for the set discussed in `Q16`. There should be a total of 730 rows in the DataFrame, which includes 676 parts and 54 spare parts. You can check the respective page on the rebrickable website to verify these results.

In [ ]:
### CODE CHECK ###

  **Q23:**  Get a list of all unique color names, sort by color name, and output as a list.  Save this as `Q23`. 

In [ ]:
### ENTER CODE HERE ###

**Q24:** How many total transparent colors are in the `colors` DataFrame?  Save this as `Q24` (as an integer).

In [ ]:
### ENTER CODE HERE ###

**Q25:**  
- Create a function called `get_part_colors` that takes as input a part number (as a string) and returns a DataFrame of all colors a part has appeared in.  
- The DataFrame should be sorted by color id  
- The index should go from 0 to n-1.

In [ ]:
### ENTER CODE HERE ###

## Optional Exercises
These exercises are optional and are not graded by CodeGrade.  They can be used as extra optional practice.

- What minifig has the most number of parts?
- What minifig shows up in the most number of sets?
- When was the first minifig included in a set?
- What set has the most number of minifigs
- What part is used the most for all sets?
- What color is used the most for all parts?
- What part / color combination is used the most for all sets?
- What part / color combination is used the least for all sets?
- Explore some of your favorite themes and which colors are used the most for those themes.
- Plot a line graph of the average number of parts per set per year.
